File: 백준/Silver/17086. 아기 상어 2/아기 상어 2.cc
#include <iostream>
#include <queue>
using namespace std;

int N, M;
int graph[51][51];
int visited[51][51] = { 0 };
queue<pair<int,int>> q;

void bfs(int r, int c) {
    q.push({r, c});

    int x, y, nx, ny;

    int dx[8] = {-1, -1, -1, 0, 0, 1, 1, 1};
    int dy[8] = {-1, 0, 1, -1, 1, -1, 0, 1};

    while (!q.empty()) {
        x = q.front().first;
        y = q.front().second;
        q.pop();

        for (int i = 0; i < 8; i++) {
            nx = x + dx[i];
            ny = y + dy[i];

            if (nx >= 0 && nx < N && ny >= 0 && ny < M) {
                if (graph[nx][ny] == 0 && (!visited[nx][ny] || visited[nx][ny] > visited[x][y] + 1)) {
                    visited[nx][ny] = visited[x][y] + 1;
                    q.push({nx, ny});
                }
            }
        }
    }
}

int main() {
    cin >> N >> M;

    for (int i = 0; i < N; i++) {
        for (int j = 0; j < M; j++) {
            cin >> graph[i][j];
        }
    }

    for (int i = 0; i < N; i++) {
        for (int j = 0; j < M; j++) {
            if (graph[i][j] == 1) {
                bfs(i, j);
            }
        }
    }

    int ans = -1;
    
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < M; j++) {
            if (ans < visited[i][j]) {
                ans = visited[i][j];
            }
        }
    }

    cout << ans;

    return 0;
}

File: 백준/Silver/17086. 아기 상어 2/아기 상어 2.py
import sys
from collections import deque
input = sys.stdin.readline

n, m = map(int, input().split())

graph = [list(map(int, input().split())) for _ in range(n)]
visited = [[0] * m for _ in range(n)]

def bfs(r, c):
    q = deque([(r, c)])

    distance = r + c

    while q:
        r, c = q.popleft()

        dx = [-1, -1, -1, 0, 0, 1, 1, 1]
        dy = [-1, 0, 1, -1, 1, -1, 0, 1]

        for i in range(8):
            nx = r + dx[i]
            ny = c + dy[i]
            if 0 <= nx < n and 0 <= ny < m:
                if graph[nx][ny] == 0 and (not visited[nx][ny] or visited[nx][ny] > visited[r][c]+1):
                    visited[nx][ny] = visited[r][c] + 1
                    q.append((nx, ny))


for i in range(n):
    for j in range(m):
        if graph[i][j] == 1:
            bfs(i, j)

ans = -1

for i in range(n):
    for j in range(m):
        if ans < visited[i][j]:
            ans = visited[i][j]
print(ans)